pkg/tracex/trace.go

```diff
@@ -0,0 +1,241 @@
+	maybeAddTagsFromPacket(e, packet)
+	t.events = append(t.events, e)
+}
+
+// OnOutgoingPacket is called when a packet is about to be sent.
+func (t *Tracer) OnOutgoingPacket(packet *model.Packet, stage int, retries int) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	stg := session.SessionNegotiationState(stage)
+	e := newEvent(handshakeEventPacketOut, stg, t.TimeNow(), t.zeroTime)
+	e.LoggedPacket = logPacket(packet, retries, model.DirectionOutgoing)
+	maybeAddTagsFromPacket(e, packet)
+	t.events = append(t.events, e)
+}
+
+// OnDroppedPacket is called whenever a packet is dropped (in/out)
+func (t *Tracer) OnDroppedPacket(direction model.Direction, stage int, packet *model.Packet) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	stg := session.SessionNegotiationState(stage)
+	e := newEvent(handshakeEventPacketDropped, stg, t.TimeNow(), t.zeroTime)
+	e.LoggedPacket = logPacket(packet, -1, direction)
+	t.events = append(t.events, e)
+}
+
+// Trace returns a structured log containing a copy of the array of [model.HandshakeEvent].
+func (t *Tracer) Trace() []*event {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return append([]*event{}, t.events...)
+}
+
+func logPacket(p *model.Packet, retries int, direction model.Direction) optional.Value[LoggedPacket] {
+	retriesValue := optional.None[int]()
+	if retries != -1 {
+		retriesValue = optional.Some(retries)
+	}
+	logged := LoggedPacket{
+		Opcode:      p.Opcode,
+		ID:          p.ID,
+		ACKs:        optional.None[[]model.PacketID](),
+		Direction:   direction,
+		PayloadSize: len(p.Payload),
+		Retries:     retriesValue,
+	}
+	if len(p.ACKs) != 0 {
+		logged.ACKs = optional.Some(p.ACKs)
+	}
+	return optional.Some(logged)
+}
+
+// LoggedPacket tracks metadata about a packet useful to build traces.
+type LoggedPacket struct {
+	Direction model.Direction
+
+	// the only fields of the packet we want to log.
+	Opcode model.Opcode
+	ID     model.PacketID
+	ACKs   optional.Value[[]model.PacketID]
+
+	// PayloadSize is the size of the payload in bytes
+	PayloadSize int
+
+	// Retries keeps track of packet retransmission (only for outgoing packets).
+	Retries optional.Value[int]
+}
+
+// MarshalJSON implements json.Marshaler.
+func (lp LoggedPacket) MarshalJSON() ([]byte, error) {
+	j := struct {
+		Direction   string                           `json:"operation"`
+		Opcode      string                           `json:"opcode"`
+		ID          model.PacketID                   `json:"id"`
+		ACKs        optional.Value[[]model.PacketID] `json:"acks"`
+		PayloadSize int                              `json:"payload_size"`
+		Retries     optional.Value[int]              `json:"send_attempts"`
+	}{
+		Direction:   dirMap[lp.Direction.String()],
+		Opcode:      lp.Opcode.String(),
+		ID:          lp.ID,
+		ACKs:        lp.ACKs,
+		PayloadSize: lp.PayloadSize,
+		Retries:     lp.Retries,
+	}
+	return json.Marshal(j)
+}
+
+var dirMap = map[string]string{
+	"recv": "read",
+	"send": "write",
+}
```

Comment:
Why and where do we need to perform this transformation?

Reply:
in retrospective it's kind of dummy; it's just because the string that came from the internal package (in model) was different from what I decided to change it to. I think I should just change the string serialization in model.Direction... and be done.